src/main/java/org/openjdk/jextract/impl/TreeMaker.java

```diff
@@ -427,7 +426,37 @@ private Declaration.Variable createVar(Cursor c, Declaration.Variable.Kind kind)
         if (c.isBitField()) throw new AssertionError("Cannot get here!");
         checkCursorAny(c, CursorKind.VarDecl, CursorKind.FieldDecl, CursorKind.ParmDecl);
         Type type = toType(c);
-        return Declaration.var(kind, CursorPosition.of(c), c.spelling(), type);
+        return withNestedTypes(Declaration.var(kind, CursorPosition.of(c), c.spelling(), type), c, false);
+    }
+
+    /*
+     * Some declarations (global vars, struct/union fields, function parameter/return types might contain
+     * inline nested struct/union/enum definitions. This method collects such definitions and
+     * attaches them to the original declaration, using an attribute.
+     */
+    private <D extends Declaration> D withNestedTypes(D d, Cursor c, boolean ignoreNestedParams) {
+        List<Declaration> nestedDefinitions = new ArrayList<>();
+        collectNestedTypes(c, nestedDefinitions, ignoreNestedParams);
+        List<Type.Declared> nestedTypes = nestedDefinitions.stream()
+                .filter(m -> m instanceof Scoped)
+                .map(s -> Type.declared((Scoped)s))
+                .toList();
+        if (!nestedTypes.isEmpty()) {
+            NestedTypes.with(d, nestedTypes);
+        }
+        return d;
+    }
+
+    private void collectNestedTypes(Cursor c, List<Declaration> nestedTypes, boolean ignoreNestedParams) {
+        c.forEach(m -> {
+            if (m.isDefinition()) {
+                if (m.kind() == CursorKind.ParmDecl && !ignoreNestedParams) {
```

Comment:
Note: if we see a param declaration, we need to recursively inspect it, as it might contain more nested declarations. This is useful for function pointer typedefs and variables whose type is a function pointer. In case of a function declaration this logic is disabled, as a function declaration will always point to its parameter declarations anyway.